clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1607,7 +1607,18 @@ class SyclKernelBodyCreator
   }
 
   bool leaveStruct(const CXXRecordDecl *, FieldDecl *FD) final {
-    const CXXRecordDecl *RD = FD->getType()->getAsCXXRecordDecl();
+
```

Comment:
Your replacement for the setting of RD should be this:
    QualType FieldTy = FD->getType();
    CXXRecordDecl *RD = FieldTy->getAsCXXRecordDecl();
    if (const ConstantArrayType *CAT =
            SemaRef.Context.getAsConstantArrayType(FieldTy))
      RD = CAT->getElementType()->getAsCXXRecordDecl();

thereby getting RD right.

Then just fall through to the existing code. That will make arrays of SYCLStream also work, which won't work with your proposed fix.

Reply:
Also, do you still need the RD == nullptr check within addStructInit after the above fix?

Reply:
Thanks for the review. I'll make the changes and upload a patch shortly. I don't think the nullptr check is needed any longer.

Reply:
I don't think qualifiers are necesary, so this whole thing could likely use: https://clang.llvm.org/doxygen/classclang_1_1Type.html#a866ff611f24817f77eb80393de8cb126

const CXXRecordDecl *RD = FD->getType()->getBaseElementTypeUnsafe()->getAsCXXRecordDecl();

Reply:
Instead of CAT->getElementType()->getAsCXXRecordDecl();? Is there an advantage of using getBaseElementTypeUnsafe()?

Reply:
Oh! That makes sense. I glanced too quickly at the definition and thought it could only be used for arrays. I'll make the change and upload new patch. Thanks!